### ISSUE TITLE
cuSPARSE Dropped support for sparse BSR

### ISSUE THREAD

**Issue by amjames** [pytorch/pytorch#113053]
### 🚀 The feature, motivation and pitch

When building PyTorch against cuda toolkit 12.2 with `-Werror` there are notices that a few places where the cuSPARSE legacy API for BSR is being used are now deprecated. 

Specifically the various info-structs that are created/used along size `cusparse<t>bsrsv2_solve` and `cusparse<t>bsrsm2_solve` now raise deprecation warnings on use. 

These APIs have been marked deprecated, but the cuSPARSE documentation indicates that the BSR layout is not supported by the generic API for sparse solve [`cusparseSp_Sv_...`](https://docs.nvidia.com/cuda/cusparse/#cusparsespsv) there is no clear migration target for the deprecated APIs. 

Looking for nvidia/cuSPARSE team to advise on the proper way to access this functionality for cuSPARSE 12.2+. 

See related: #110624

### Alternatives

A workaround would involve converting BSR to a supported format around this operation. This is not a preferred way to maintain functionality. 

### Additional context

See related: #110624

**Comment by ptrblck** [comment]
@amjames Could you forward me any models, layers, etc. using these (currently deprecated) methods, please?

**Comment by malfet** [comment]
Should this be closed as duplicate as https://github.com/pytorch/pytorch/issues/110624 (or other way around as this one is more informative than original issue)

**Comment by amjames** [comment]
> Could you forward me any models, layers, etc. using these (currently deprecated) methods, please?

 This is a valid question, I don't know of any off the top of my head, but I can track down some users of the feature from some older issues to see what they were needing it for. 

The main issue here is that we have had the feature via cuSPARSE for some time, and I find it hard to believe that the API was deprecated without a usable migration target, @ptrblck are you able to confirm with the cuSPARSE team that this deprecation was not a mistake? Or that the BSR layout support has not been added to the generic API?

In the past, deprecation of the legacy API in cuSPARSE usually goes hand-in-hand with support for the layout being added to the generic api for that OP.

**Comment by amjames** [comment]
@tvercaut Could you describe your use case for sparse triangular solve? I see you authored  #87358 related to the feature.

**Comment by tvercaut** [comment]
@amjames Is your use case question specifically about BSR storage and sparse triangular solve? I don't currently use BSR storage. I do however heavily rely on CSR storage with sparse triangular solve and more. In #87358, I reported about BSR as I only "accidentally" found out that `torch.triangular_solve` supported CSR by reading somewhere (can't recall where) that BSR support had been added for it in pytorch. I then dug into the code base to see if by any chance CSR wasn't there as well and I was lucky to find it.

Anyway, maybe the team behind [baspacho](https://github.com/facebookresearch/baspacho) could some potential use cases?

**Comment by ptrblck** [comment]
> and I find it hard to believe that the API was deprecated without a usable migration target, @ptrblck are you able to confirm with the cuSPARSE team that this deprecation was not a mistake?

Yes, it's a planned deprecation unless we find a valid use case for this API besides the current status where it was exposed because it was available. 

So far, we could not identify a single use case, so let me know if you have one.

**Comment by drisspg** [comment]
I typically build from source with WERROR=1, since this generally saves me a round tip to CI/CD. I currently can't build from source with this setting on cuda 12.2. 

Is there a work around?

**Comment by nWEIdia** [comment]
This issue is now blocking !121956 , e.g. failures in https://hud.pytorch.org/pr/pytorch/pytorch/121956#25010023465

**Comment by drisspg** [comment]
I was annoyed by this, since it won't let me set WERROR = 1 locally (using 12.3).  

One option is to set  add_compile_options(-Wno-deprecated-declarations). But this feels bad and should likely only be done on these specific Sparsity files. 

I don't think that is the correct solution though, and is merely a workaround.

**Comment by amjames** [comment]
Since it seems that the deprecation of this API without a migration target was intentional (see @ptrblck 's comment above) we will have to work around it on our side. Here is a potential plan. 

Short Term: We can add `-Wno-deprecated-declarations` to the appropriate files in our build, and add deprecation warnings on the appropriate ops . Eventually when the APIs are actually removed we will have to disable compilation of those code blocks. Our wrapper for the BSR solve descriptor is already conditionally defined on `#if AT_USE_HIPSPARSE_TRIANGULAR_SOLVE()`  we could change that macro out for one which also checks the cusparse version. 

If that short term solution surfaces users who are relying on the functionality we can plan to do something to preserve the feature after cusparse removes the deprecated APIs. I have looked into other sparse solver libraries we could depend on an for block sparse layouts there are not really any options, so we would probably be implementing it ourselves. If any of those use-cases are compelling maybe cuSPARSE will decide to keep the API.

If we don't turn up any users who are going to miss the feature we probably don't need to implement it and take on the maintenance IMO.

**Comment by nWEIdia** [comment]
cc @Aidyn-A and @eqy

**Comment by cpuhrsch** [comment]
@ptrblck - We recently published https://pytorch.org/blog/speeding-up-vits/ that uses BSR kernels generated via Triton. Would be nice to see cuSPARSE improve upon this if there's interest.

**Comment by eqy** [comment]
Would these just be using solver kernels (deprecated) or BSRMM (apparently still suported)?
